contracts/core/Dispatcher.sol

```diff
@@ -474,6 +526,15 @@ contract Dispatcher is IbcDispatcher, IbcEventsEmitter, Ownable, Ibc {
         );
     }
 
+    // Returns the result of the call if no revert, otherwise returns the error if thrown.
+    function _try_catch(address portAddress, bytes memory args) internal returns (bool success, bytes memory message) {
+        if (!Address.isContract(portAddress)) {
+            return (false, bytes("call to non-contract"));
+        }
+        // Only call if we are sure portAddress is a contract
+        (success, message) = portAddress.call(args);
+    }
```

Comment:
this is not doing a try/catch. It's only checking if the address is a contract. I thought the whole point of the try/catch was to prevent txs being reverted due to dapps. If the dapp callback decides to revert or assert the tx will be reverted.

Reply:
> If the dapp callback decides to revert or assert the tx will be reverted.
If the dapp callback reverts, the low level call will just return false for success, but not revert our tx. Which is what we wanted? 
- we wanted to log an error, but not revert the tx to the dispatcher in the case of
    -  trying a call to a non-contract address
    -  trying a call to a contract which doesn't support the callback method
    -  trying a call a contract which does support the callback but which reverts the call
- we wanted to succeed all other calls and log an event (dependent on the handshake 2-step) in the case

with the current code, what happens: 
- A call to a non-contract address will return false for `success` and `call to non-contract` as the data in the `try_catch` 
- A call to a contract which doesn't support the callback interface will return false for `success` and an empty error in the `try_catch` method
- A call to a contract which does support the callback but throws in the callback will return false for `success` and the error that the contract reverted with, which will both be returned in this method
- A call to a successful contract callback will return true for success and the data that the callback returns, if any

This is what we wanted, yes? We can also verify this is how the code is working through the test suite - if you think there is another test case that is missing let me know and I can add it

Reply:
FYI, external calls to non-contract addresses don't revert themselves; since non-contract addresses don't have any code, they can't revert. So a low-level call will always return success for a non-contract address, which is why I have that logic in the code the way it is

Reply:
Our previous implementation of solidity's try/catch would have thrown errors in the case of call to a non-contract address or to a contract which doesn't have the method we are trying to call, and I don't think there is a way to account for that using solidity's try/catch because solidity's try catch only applies to a specific external call, and in these cases, the revert doesn't happen in the external call itself, but instead it happens within our dispatcher contract so the tx will still revert on the dispatcher level despite the use of solidity's try/catch 

I think the only way around this is to use low-level calls

Reply:
ah that makes sense! All cases are covered then and indeed we haven't changed the approach. Thanks for the thorough explanation.